### ISSUE TITLE
Feature Request - Add Cell Number

### ISSUE THREAD

**Issue by paris0120** [jupyterlab/jupyterlab#5905]
I believe it will be very useful if each cell can have a number before them (like line number in many ide)

**Comment by jasongrout** [comment]
Thanks for the suggestion. We'd have to be careful about how balance this with the execution number that already appears by cells.

To focus a discussion and the design decision and tradeoffs, what are the main usecases you have for a cell number?

**Comment by paris0120** [comment]
For reference when I talk with someone and when I can get back quickly from one cell to another, for example change a previous step then rerun a result later. Easier to location cells with a number. You can put the cell number before the step number like #5 [12]

**Comment by bollwyvl** [comment]
I feel like cosmetic numbers are a bit of a band-aid.

For the commenting/metadata work, cell ids will become very useful, so
you'd want this to be predictable. At present, there is an nbformat cell
metadata field, name, for adding _something_ unique to cells, though I
don't know if it is validated by any clients.

I'd like to see this used, and become an anchor source. By default, having
it populate with a namespaced, increasing integer, e.g jp:1, would be
great. Cell copy paste would have to be aware of it, and be able to assign
fresh numbers. You could left pad with zeros, I suppose. Luckily there's a
package for that!

Additionally, having an "autorenumber" command would be good, as it would
help in version control workflows. If a user actually names their cell
(which is good!), those would be off limits, but we'd need a plan for
fixing those numbers as well.

Then the UI: yeah, introducing a second number would be rough next to
prompt/status. Perhaps, in keeping with the markdown headers, it could go
on the right, but perhaps still inside/above the cell, and be editable on
double click (gateway to metadata), while a single click would put that
anchor in the URL. And the whole shooting match could be toggleable, "show
cell names".

**Comment by jukworks** [comment]
I totally agree that it makes UI more confusing because execution numbers already exist.
But when I run nbQA(https://github.com/nbQA-dev/nbQA), it gives me error/warning positions like 'cell_{N}'. I can quickly access the cell if Jupyter shows cell numbers. @jasongrout
Thanks.

**Comment by krassowski** [comment]
What do folks think about adding a cell number (index) in the hover over the execution counter? Currently it shows nothing, but I imagine we could have cell execution, index and ID show up on hover:

<img width="698" height="290" alt="Image" src="https://github.com/user-attachments/assets/02d94ab6-1d1d-4989-9fcb-12ed36c69c30" />

Maybe even combined with some hints on drag-and-drop capabilities:

<img width="870" height="300" alt="Image" src="https://github.com/user-attachments/assets/6a1b3a03-f4ee-4888-9ad7-abc7947252f0" />

I agree that we should not display another number in UI at all times, it would take up real estate and be quite confusing.

**Comment by jasongrout** [comment]
Interesting, Mike, thanks for proposing a solution here!

I would love for an index number to be available somewhere in the ux. I saw how useful they were in Databricks notebooks, for example, to do exactly the kind of quick referencing of specific cells in a notebook mentioned above. In Databricks, the cell index is noted as some small text in the header of a cell, and you can add a title to a cell index number to make it easier to reference a cell with a semantic name. You can click on that index number to edit the title in-place.

<img width="847" height="421" alt="Image" src="https://github.com/user-attachments/assets/566f47e5-925e-4ccd-b0d7-6ee47c7c9561" />

As for exposing these in a hover, I have a few concerns:

1. Hovers in general aren't accessible, I think
2. Having large hovers obscures notebook content. In this case, we'd be introducing a lot of hover targets, so my guess is that in practice you'll fairly often see these hovers pop up in normal usage when you'd rather not see them.
3. Having them in an ephemeral UI element makes it tedious to quickly locate a cell based on the index number, which is the prominent usecase mentioned above. In other words, I can't just scroll through a notebook and quickly locate the right cell.

**Comment by krassowski** [comment]
> 1. Hovers in general aren't accessible, I think

`title` attributes have two categories of issues:
- they are harde to access on devices without a mouse-like pointer. This is also the case for the file browser (for which there is an exploration of details popup which would help with it, https://github.com/jupyterlab/jupyterlab/pull/18289). So, if we went for `title` we should probably _also_ find a different way to expose it.
- they can be read as-is by screen readers so they need to be carefully tested with relation to other attributes (like combined with aria-label) to avoid spamming the reader

Other DOM elements which appear on hover are indeed inaccessible for different reasons, but the `title` is relatively accessible if these issues are taken care of.


> 2. Having large hovers obscures notebook content. In this case, we'd be introducing a lot of hover targets, so my guess is that in practice you'll fairly often see these hovers pop up in normal usage when you'd rather not see them.

Fair, I agree. I guess we could have it smaller like `executed 1st, index 3`. Adding cell ID would necessarily make it much larger.


> 3. Having them in an ephemeral UI element makes it tedious to quickly locate a cell based on the index number, which is the prominent usecase mentioned above. In other words, I can't just scroll through a notebook and quickly locate the right cell.

What if we had the cell index of the active cell on the status bar? And maybe also the range if multiple cells are selected?

<img width="898" height="540" alt="Image" src="https://github.com/user-attachments/assets/dca85f18-0bc1-49f0-a9d6-3765a6d8c017" />

- I often wonder how many cells are in the notebook, this would help with such use case too.
- The line/column item has `Go to line number…` we could make this one `Go to cell…` with dialog options for:
   - https://github.com/jupyterlab/jupyterlab/issues/18195
   - https://github.com/jupyterlab/jupyterlab/issues/9786
- Showing the range of selected cells is a common pattern for spreadsheet editors like Excel, and e.g. [`jupyterlab-spreadsheet-editor](https://github.com/jupyterlab-contrib/jupyterlab-spreadsheet-editor) does that. They do not need to show row/column in statusbar because they have the row and column numbers displayed at all times

<img width="683" height="488" alt="Image" src="https://github.com/user-attachments/assets/1cc64f3e-65f0-4c4a-856c-8f9cb9a810a5" />

**Comment by bollwyvl** [comment]
Yep, much like with the file browser, extension- and ultimately user-controllable grids are magic.

We presently have a bunch of special tooling for 3 (meta)data kinds, which can be (or are forced, if on a small screen) to be either rows or columns:
- prompt
- source 
- outputs with default renderer

In the general case, this could be described as a schema of table options, which could optionally add new projections of data/metadata like:
- a cell number/identifier column with anchor links
- cell actions
- comments
- VCS information
- additional output renderers
- tags
- translations

These could entirely be a user concern, or potentially be stored in notebook metadata.
